backlinks/obs_filter.js

```diff
@@ -0,0 +1,80 @@
+var nest = require('depnest')
+var pull = require('pull-stream')
+var MutantPullReduce = require('mutant-pull-reduce')
+var Value = require('mutant/value')
+
+exports.needs = nest({
+  'sbot.pull.backlinks': 'first'
+})
+
+exports.gives = nest("backlinks.filter.obs", true)
+
+/**
+ * sbot.filter.obs returns an observable list of messages that link
+ * back to the message with the given message ID (@id). Only messages that
+ * pass the filter function are added to the list.
+ *
+ * When a message arrives, if passing it to the filter function (@filterFn)
+ * does not result in it returning 'true' the message is not added to the
+ * observable list.
+ *
+ * A 'sync' observable property is also added to the returned observable
+ * which is 'true' when all previously seen messages are caught up with.
+ *
+ * Note: Unlike backlinks.obs.for this does not cache the observable for
+ * callers that supply the same arguments.
+ */
+exports.create = function(api) {
+  function pullFilterReduceObs(id, filterFn) {
+    if (!id || typeof(id) !== "string") {
+      throw new Error("id must be a string.")
+    }
+
+    if (!filterFn || typeof(filterFn) !== "function") {
+      throw new Error("filterFn must be a function.")
+    }
+
+    var sync = Value(false)
+
+    var msgBacklinks = api.sbot.pull.backlinks({
+      query: [{
+        $filter: {
+          dest: id
+        }
+      }],
+      live: true
+    })
+
+    var msgFilter = pull.filter((msg) => {
+      // We do not include the 'sync' message which indicates that any
+      // further messages are newly gossiped messages in the list of messages.
+      if (msg.sync) {
```

Comment:
`mutant-pull-reduce` [actually already handles the sync message](https://github.com/mmckegg/mutant-pull-reduce/blob/master/index.js#L39), and it exposes an `obs.sync`. So this isn't really needed.

At the moment, it defaults to doing this, but it should really be an option. I suggest that if you use this, you should add `{sync: true}` to the `MutantPullReduce` options.

Reply:
Huh. That's weird. I originally added this code because in ssb-chess I already have a watch on obs.sync and it wasn't getting fired (therefore the games weren't getting rendered), so I assumed I had to take care of the sync myself.

I just removed this stuff, and then set a breakpoint on https://github.com/mmckegg/mutant-pull-reduce/blob/master/index.js#L39 and I notice that it's never getting set to true (i.e. that line is never executed.) To make sure I was breakpointing the right file, I added a breakpoint to the line that pushes the item and that did work.

I was wondering if you could help me debug this?

I'll push my updates to this pull request (including setting the 'sync' observable manually) for now and we can try and fix the bug in mutant-pull-reduce before merging this.